dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,174 @@
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `FailedRequestClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = FailedRequestClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
+    Client? client,
+    Hub? hub,
+  })  : _hub = hub ?? HubAdapter(),
+        _client = client ?? Client();
+
+  final Client _client;
+  final Hub _hub;
+
+  /// Configures up to which size request bodies should be included in events
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<int> failedRequestStatusCodes;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    try {
+      final response = await _client.send(request);
+
+      if (failedRequestStatusCodes.contains(response.statusCode)) {
+        await _captureException(request: request);
+      }
+
+      return response;
+    } catch (exception, stackTrace) {
+      await _captureException(
+        exception: exception,
+        stackTrace: stackTrace,
+        request: request,
+      );
+
+      rethrow;
+    }
+  }
+
+  @override
+  void close() {
+    // See https://github.com/getsentry/sentry-dart/pull/226#discussion_r536984785
+    _client.close();
+  }
+
+  // See https://develop.sentry.dev/sdk/event-payloads/request/
+  Future<void> _captureException({
+    Object? exception,
+    StackTrace? stackTrace,
+    required BaseRequest request,
+  }) {
+    // As far as I can tell there's no way to get the uri without the query part
+    // so we replace it with an empty string.
+    final urlWithoutQuery = request.url.replace(query: '').toString();
+
+    final query = request.url.query.isEmpty ? null : request.url.query;
+
+    final sentryRequest = SentryRequest(
+      method: request.method,
+      headers: request.headers,
+      url: urlWithoutQuery,
+      queryString: query,
+      data: _getDataFromRequest(request),
+      other: {
+        'Content-Length': request.contentLength.toString(),
+      },
+    );
+    final event = SentryEvent(
+      throwable: exception,
+      request: sentryRequest,
+      level: SentryLevel.error,
+    );
```

Comment:
let's add a mechanism to it, like we do with other type of events

https://github.com/getsentry/sentry-dart/blob/main/dart/lib/src/default_integrations.dart#L31-L37

so people could filter by the mechanism